sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/sqs/SqsUnboundedReader.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.aws.sqs;
+
+import com.amazonaws.services.sqs.AmazonSQS;
+import com.amazonaws.services.sqs.AmazonSQSClientBuilder;
+import com.amazonaws.services.sqs.model.Message;
+import com.amazonaws.services.sqs.model.ReceiveMessageRequest;
+import com.amazonaws.services.sqs.model.ReceiveMessageResult;
+import com.google.common.collect.Lists;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayDeque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+import java.util.Set;
+import org.apache.beam.sdk.io.UnboundedSource;
+import org.apache.beam.sdk.io.UnboundedSource.CheckpointMark;
+import org.joda.time.Instant;
+
+class SqsUnboundedReader extends UnboundedSource.UnboundedReader<Message> {
+
+  public static final int MAX_NUMBER_OF_MESSAGES = 10;
+  private final SqsUnboundedSource source;
+  private final AmazonSQS sqs;
+
+  private Message current;
+  private final Queue<Message> messagesNotYetRead;
+  private Set<String> receiptHandlesToDelete;
+
+  public SqsUnboundedReader(SqsUnboundedSource source, SqsCheckpointMark sqsCheckpointMark) {
+    this.source = source;
+    this.current = null;
+
+    this.messagesNotYetRead = new ArrayDeque<>();
+    receiptHandlesToDelete = new HashSet<>();
+
+    final SqsConfiguration sqsConfiguration = source.getSqsConfiguration();
+    sqs =
+        AmazonSQSClientBuilder.standard()
+            .withClientConfiguration(sqsConfiguration.getClientConfiguration())
+            .withCredentials(sqsConfiguration.getAwsCredentialsProvider())
+            .withRegion(sqsConfiguration.getAwsRegion())
+            .build();
+
+    if (sqsCheckpointMark != null) {
+      if (sqsCheckpointMark.getReceiptHandlesToDelete() != null) {
+        receiptHandlesToDelete.addAll(sqsCheckpointMark.getReceiptHandlesToDelete());
+      }
+    }
+  }
+
+  @Override
+  public Instant getWatermark() {
+    return Instant.now();
```

Comment:
This ignores the original message timestamps so probably worth to use an approach similar to what [Jms does (saving the oldest timestamp](https://github.com/apache/beam/blob/f2ae73be3d3d1cb7c3d111505f832aed44bc92eb/sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsCheckpointMark.java).

Notice however that if the queue is not standard (the FIFO case) [just updating the watermark to the message timestamp](https://github.com/apache/beam/blob/f2ae73be3d3d1cb7c3d111505f832aed44bc92eb/sdks/java/io/amqp/src/main/java/org/apache/beam/sdk/io/amqp/AmqpIO.java#L308-L309) could be enough.

Reply:
I am still wondering if it is worth to have different strategies depending if it uses a Standard or FIFO queue. But it seems the standard case should cover the timestamp approximately well for the FIFO one. Any ideas? (of course this can be subject of other JIRA)

Reply:
Perhaps I misread the javadoc comments on the base class...

"For a source that does not have natural timestamps, timestamps can be set to the time of reading, in which case the watermark is the current clock time."

https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java#L217-L219